text/0126-introduce-xcq.md

```diff
@@ -0,0 +1,354 @@
+        fn metadata() -> Vec<u8>;
+    }
+}
+type XcqResult =  Result<XcqResponse, XcqError>;
+type XcqResponse = Vec<u8>;
+enum XcqError {
+    Custom(String),
+}
+```
+
+**Example Metadata (before SCALE-encoded)**
+
+```rust
+pub struct Metadata {
+    pub types: PortableRegistry,
+    pub extensions: Vec<ExtensionMetadata<PortableForm>>,
+}
+```
+
+### XCQ Executor
+
+An XCQ executor is a runtime module that executes XCQ queries.
+It has a core method `execute` that takes a PolkaVM program binary,
+method name of the exported functions in the PolkaVM program, input arguments, and weight limit that the PolkaVM program can consume.
+
+```rust
+pub fn execute(
+    &mut self,
+    raw_blob: &[u8],
+    method: &str,
+    input: &[u8],
+    weight_limit: u64,
+) -> Result<Vec<u8>, XcqExecutorError> {...}
+```
+
+### XCQ Extension
+
+An extension-based design is essential for several reasons:
+
+- Different chains may have different data types for semantically similar queries, making it challenging to standardize function calls across them.
+An extension-based design with optional associated types allows these diverse data types to be specified and utilized effectively.
+- Function calls distributed across various pallets can be amalgamated into a single extension, simplifying the development process and ensuring a more cohesive and maintainable codebase.
+- New functionalities can be added without upgrading the core part of the XCQ.
+- Ensure the core part is in a minimal scope.
+
+Essential components of an XCQ extension system include:
+
+- A hash-based extension id generation mechanism for addressing and versioning . The hash value derives from the extension name and its method sets. Any update to an extension is treated as a new extension.
+
+- `decl_extension` macro: Defines an extension as a Rust trait with optional associated types.
+
+**Example usage**:
+
+```rust
+use xcq_extension::decl_extension;
+
+pub trait Config {
+    type AssetId: Codec;
+    type AccountId: Codec;
+    type Balance: Codec;
+}
+decl_extension! {
+    pub trait ExtensionFungibles {
+        type Config: Config;
+        fn total_supply(asset: <Self::Config as Config>::AssetId) -> <Self::Config as Config>::Balance;
+        fn balance(asset: <Self::Config as Config>::AssetId, who: <Self::Config as Config>::AccountId) -> <Self::Config as Config>::Balance;
+    }
+}
+```
+
+- `impl_extensions` macro: Generates extension implementations and extension-level metadata.
+
+**Example Usage**:
+
+```rust
+// ExtensionImpl is an aggregate struct to impl different extensions
+struct ExtensionImpl;
+impl extension_fungibles::Config for ExtensionImpl {
+    type AssetId = u32;
+    type AccountId = [u8; 32];
+    type Balance = u64;
+}
+impl_extensions! {
+    impl extension_core::ExtensionCore for ExtensionImpl {
+        type Config = ExtensionImpl;
+        fn has_extension(id: <Self::Config as extension_core::Config>::ExtensionId) -> bool {
+            matches!(id, 0 | 1)
```

Comment:
I thought extension ids were hashes of the content? The extension decides its id?

Reply:
When I read the RFC, I understood the `ExtensionCore` as a "special" "foundational" extension. It seems it will have some generic implementation where the extension IDs will indeed be content-addressed.

Is that so? Is this code just an example of "some return value"?

Reply:
this is just a dummy implementation. we should leave it out in the example